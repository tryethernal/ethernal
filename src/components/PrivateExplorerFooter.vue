<template>
    <v-footer class="footer">
        <v-container class="d-flex flex-column py-4">
            <!-- Links Columns Row -->
            <div class="d-flex flex-column flex-sm-row ga-md-8 ga-sm-1">
                <div class="d-flex flex-column mb-8">
                    <span class="text-body-2">
                        Block explorer & data analytics platform for evm chains.
                    </span>
                    <v-btn
                        @click="addToMetaMask"
                        variant="outlined"
                        size="small"
                        class="mt-2 px-2"
                        max-width="180px"
                        :title="'Add to wallet'">
                            <img :src="MetamaskIcon" class="metamask-icon mr-2" width="20" height="20" alt="MetaMask" />
                            <span class="text-body-2">Add to wallet</span>
                    </v-btn>
                </div>

                <div class="d-flex flex-wrap">
                    <div class="d-flex flex-column ga-2 footer-links-section">
                        <span class="text-subtitle-2">Resources</span>
                        <a class="text-decoration-none text-body-2" v-for="link in resources" :key="link.name" :href="link.url" target="_blank">
                            {{ link.name }}
                        </a>
                    </div>

                    <div class="d-flex flex-column ga-2 footer-links-section">
                        <span class="text-subtitle-2">Community</span>
                        <a class="text-decoration-none text-body-2" v-for="link in community" :key="link.name" :href="link.url" target="_blank">
                            {{ link.name }}
                        </a>
                    </div>

                    <div class="d-flex flex-column ga-2 footer-links-section">
                        <span class="text-subtitle-2">Company</span>
                        <a class="text-decoration-none text-body-2" v-for="link in company" :key="link.name" :href="link.url" target="_blank">
                            {{ link.name }}
                        </a>
                    </div>
                </div>
            </div>

            <v-divider class="my-6"></v-divider>

            <!-- Copyright and Donation Row -->
            <div class="d-flex justify-space-between align-center">
                <div class="text-caption">
                    Powered by <a :href="`https://tryethernal.com?ref=${host}`" target="_blank">Ethernal</a> | v{{ envStore.version }} | Made with üç∑ in France
                </div>
            </div>
        </v-container>
    </v-footer>
</template>

<script setup>
import { useEnvStore } from '../stores/env';
import { useCurrentWorkspaceStore } from '../stores/currentWorkspace';
import MetamaskIcon from '../assets/metamask-fox.svg';

const envStore = useEnvStore();
const host = document.location.host;
const currentWorkspaceStore = useCurrentWorkspaceStore();

const resources = [
    {
        name: 'Documentation',
        url: 'https://doc.tryethernal.com'
    },
    {
        name: 'Github',
        url: 'https://github.com/tryethernal'
    }
];

const community = [
    {
        name: 'Twitter / X',
        url: 'https://x.com/tryethernal'
    },
    {
        name: 'Discord',
        url: 'https://discord.gg/jEAprf45jj'
    },
    {
        name: 'Telegram',
        url: 'https://t.me/antoinedc'
    }
];

const company = [
    {
        name: 'Home',
        url: 'https://tryethernal.com'
    },
    {
        name: 'Contact Us',
        url: 'https://tryethernal.com/contact-us'
    },
    {
        name: 'Terms of Service',
        url: 'https://tryethernal.com/terms'
    },
    {
        name: 'Privacy Policy',
        url: 'https://tryethernal.com/privacy'
    }
];

async function addToMetaMask() {
    try {
        if (!window.ethereum)
            return;

        await window.ethereum.request({
            method: 'wallet_addEthereumChain',
            params: [{
                chainId: currentWorkspaceStore.networkId,
                chainName: currentWorkspaceStore.name,
                nativeCurrency: {
                    name: currentWorkspaceStore.chain.token,
                    symbol: currentWorkspaceStore.chain.token,
                    decimals: 18
                },
                rpcUrls: [currentWorkspaceStore.rpcServer],
                blockExplorerUrls: [`https://app.${envStore.mainDomain}`]
            }]
        });
    } catch (error) {
        console.error('Failed to add network to MetaMask:', error);
    }
}
</script>

<style scoped>
.footer {
    border-top: 1px solid rgba(var(--v-border-color), var(--v-border-opacity));
}

.metamask-icon {
    vertical-align: middle;
}

.ga-8 {
    gap: 8px;
}

.ga-2 {
    gap: 8px;
}

.footer-links-section {
    min-width: 150px;
    margin-right: 24px;
}

@media (max-width: 600px) {
    .footer-links-section {
        min-width: calc(50% - 12px);
        margin-right: 0;
    }
}

@media (max-width: 400px) {
    .footer-links-section {
        min-width: 100%;
    }
}
</style> 