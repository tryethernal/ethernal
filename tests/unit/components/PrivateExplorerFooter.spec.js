import PrivateExplorerFooter from '@/components/PrivateExplorerFooter.vue';
import { createTestingPinia } from '@pinia/testing';

describe('PublicExplorerFooter.vue', () => {
    const host = 'test.ethernal.com';
    const mockEnvStore = {
        version: '1.2.3',
        mainDomain: 'ethernal.com'
    };
    const mockCurrentWorkspaceStore = {
        networkId: '0x1',
        name: 'Test Network',
        chain: {
            token: 'ETH'
        },
        rpcServer: 'https://test.rpc'
    };

    beforeEach(() => {
        // Mock document.location
        Object.defineProperty(window, 'location', {
            value: { host },
            writable: true
        });
    });

    it('Should show the component', async () => {
        const wrapper = mount(PrivateExplorerFooter, {
            global: {
                plugins: [
                    createTestingPinia({
                        initialState: {
                            env: mockEnvStore,
                            currentWorkspace: mockCurrentWorkspaceStore
                        }
                    })
                ]
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should try to add network to metamask when clicking the button', async () => {
        const mockRequest = vi.fn();
        global.window.ethereum = {
            request: mockRequest
        };

        const wrapper = mount(PrivateExplorerFooter, {
            global: {
                plugins: [
                    createTestingPinia({
                        initialState: {
                            env: mockEnvStore,
                            currentWorkspace: mockCurrentWorkspaceStore
                        }
                    })
                ]
            }
        });
        await flushPromises();

        await wrapper.find('button').trigger('click');

        expect(mockRequest).toHaveBeenCalledWith({
            method: 'wallet_addEthereumChain',
            params: [{
                chainId: mockCurrentWorkspaceStore.networkId,
                chainName: mockCurrentWorkspaceStore.name,
                nativeCurrency: {
                    name: mockCurrentWorkspaceStore.chain.token,
                    symbol: mockCurrentWorkspaceStore.chain.token,
                    decimals: 18
                },
                rpcUrls: [mockCurrentWorkspaceStore.rpcServer],
                blockExplorerUrls: [`https://app.${mockEnvStore.mainDomain}`]
            }]
        });
    });

    it('Should not try to add network to metamask when ethereum is not available', async () => {
        global.window.ethereum = undefined;
        const consoleSpy = vi.spyOn(console, 'error');

        const wrapper = mount(PrivateExplorerFooter, {
            global: {
                plugins: [
                    createTestingPinia({
                        initialState: {
                            env: mockEnvStore,
                            currentWorkspace: mockCurrentWorkspaceStore
                        }
                    })
                ]
            }
        });
        await flushPromises();

        await wrapper.find('button').trigger('click');

        expect(consoleSpy).not.toHaveBeenCalled();
    });

    it('Should handle errors when adding network', async () => {
        const mockRequest = vi.fn().mockRejectedValue(new Error('Failed'));
        const consoleSpy = vi.spyOn(console, 'error');
        global.window.ethereum = {
            request: mockRequest
        };

        const wrapper = mount(PrivateExplorerFooter, {
            global: {
                plugins: [
                    createTestingPinia({
                        initialState: {
                            env: mockEnvStore,
                            currentWorkspace: mockCurrentWorkspaceStore
                        }
                    })
                ]
            }
        });
        await flushPromises();

        await wrapper.find('button').trigger('click');

        expect(consoleSpy).toHaveBeenCalledWith('Failed to add network to MetaMask:', expect.any(Error));
    });
}); 