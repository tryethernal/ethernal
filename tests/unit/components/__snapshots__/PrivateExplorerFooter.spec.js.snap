// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PublicExplorerFooter.vue > Should show the component 1`] = `
"<footer data-v-eccf9220="" class="v-footer v-theme--light footer" style="height: auto;">
  <div data-v-eccf9220="" class="v-container v-locale--is-ltr d-flex flex-column py-4">
    <!-- Links Columns Row -->
    <div data-v-eccf9220="" class="d-flex flex-column flex-sm-row ga-md-8 ga-sm-1">
      <div data-v-eccf9220="" class="d-flex flex-column mb-8"><span data-v-eccf9220="" class="text-body-2"> Block explorer &amp; data analytics platform for evm chains. </span><button data-v-eccf9220="" type="button" class="v-btn v-theme--light v-btn--density-default v-btn--size-small v-btn--variant-outlined mt-2 px-2" style="max-width: 180px;" title="Add to wallet"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
          <!----><span class="v-btn__content" data-no-activator=""><img data-v-eccf9220="" src="/src/assets/metamask-fox.svg" class="metamask-icon mr-2" width="20" height="20" alt="MetaMask"><span data-v-eccf9220="" class="text-body-2">Add to wallet</span></span>
          <!---->
          <!---->
        </button></div>
      <div data-v-eccf9220="" class="d-flex flex-wrap">
        <div data-v-eccf9220="" class="d-flex flex-column ga-2 footer-links-section"><span data-v-eccf9220="" class="text-subtitle-2">Resources</span><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://doc.tryethernal.com" target="_blank">Documentation</a><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://github.com/tryethernal" target="_blank">Github</a></div>
        <div data-v-eccf9220="" class="d-flex flex-column ga-2 footer-links-section"><span data-v-eccf9220="" class="text-subtitle-2">Community</span><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://x.com/tryethernal" target="_blank">Twitter / X</a><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://discord.gg/jEAprf45jj" target="_blank">Discord</a><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://t.me/antoinedc" target="_blank">Telegram</a></div>
        <div data-v-eccf9220="" class="d-flex flex-column ga-2 footer-links-section"><span data-v-eccf9220="" class="text-subtitle-2">Company</span><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://tryethernal.com" target="_blank">Home</a><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://tryethernal.com/contact-us" target="_blank">Contact Us</a><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://tryethernal.com/terms" target="_blank">Terms of Service</a><a data-v-eccf9220="" class="text-decoration-none text-body-2" href="https://tryethernal.com/privacy" target="_blank">Privacy Policy</a></div>
      </div>
    </div>
    <hr data-v-eccf9220="" class="v-divider v-theme--light my-6" aria-orientation="horizontal" role="separator"><!-- Copyright and Donation Row -->
    <div data-v-eccf9220="" class="d-flex justify-space-between align-center">
      <div data-v-eccf9220="" class="text-caption"> Powered by <a data-v-eccf9220="" href="https://tryethernal.com?ref=localhost:3000" target="_blank">Ethernal</a> | v1.2.3 | Made with üç∑ in France </div>
    </div>
  </div>
</footer>"
`;
